_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+saying that the position of `'o` in the context of `Output`'s declaration
+corresponds with the position of `'b` in `Output<'b>` in the method. Next, we
+see the `data: &'o Opaque` field, and use the mapping to determine that it's
+actually `&'b Opaque` in the context of the method. We then remark that
+`out.data` relies on `'b`, see that there's no more fields to explore in the
+field traversal, and finish.
+
+Once we know which lifetimes to look out for (just `'b`), we traverse the types
+of the
+inputs, which consists of just `Input<'a>`. Since `Input` is a non-opaque struct,
+we look in the declaration of `Input` and store a mapping saying that the
+position of `'i` in the context of `Input`'s declaration corresponds with the
+position of `'a` in `Input<'a>` in the method. Inside of `Input`s declaration
+we see the `data: &'i Opaque` field, so we map the lifetime `'i` back to `'a`,
+and then perform a depth-first search in the lifetime graph to find all shorter
+lifetimes.  This search results in finding `'b`. After confirming that `'b`
+appears in the output, we remark that `input.data` is one of the input fields
+that lives at least `'b`.
+
+In code generation, we put these pieces of information together to determine
+that `out.data` must be outlived by `input.data`, which is exactly what we
+were aiming for.
+
+# Results
+That's a lot of details, and unfortunately it's not 100% integrated yet.
+Thus far, the lifetime coercion detection using depth-first search
+is entire implemented, but tracking borrows on a field-by-field basis is only
+implemented in the JavaScript backend. Aside from not understanding the full
+scope of the problem before starting, the biggest challenge was that Diplomat's
+abstract representation was not designed with lifetimes in mind, and also
+wasn't condusive to writing robust code.
+
+For example, traversing down structs recursively and maintaining a mapping from
+lifetimes in a declaration to lifetimes in the method was bolted on at the end
+after I'd already written a significant amount of code. Additionally, Diplomat's
+model allows for a bunch of illegal states like an opaque struct passed by value.
+Obviously there are validity checks, the loose type system eventaully pollutes
+backends with `.unwrap()` and `unreachable!()` statements. On top of this, the
+existing JavaScript code generation was really challenging to understand and
+work with, and I ended up spending a week rewriting it to be more modular.
+
+# Next Steps
+Diplomat is relatively a new project, and we're still making a lot of decisions
+that we're unsure about. Since I was adding such a large feature to Diplomat's
+core model, a lot of these older design decisions came back to bite me, and
+I began to consider how I would have implemented Diplomat knowing what I know
+now. Additionally, Manish and I had been discussing how to more concretely
+define Diplomat's type model, particularly around input and output types which
```

Comment:
tbh "input and output types" isn't particularly relevant for _this_ audience: input and output types are the thing about AST that is particularly messy, but it's not the only thing

The way I'd frame this is that we wanted a better type model, and you talk about what that means below.